server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -146,13 +146,19 @@ public Void visitFunction(Function function, NullabilityContext context) {
                 // datatype is an object. There we need to exclude null values to not match
                 // empty objects on the query
                 var a = function.arguments().get(0);
-                var b = function.arguments().get(1);
-                if (a instanceof Reference ref && b instanceof Literal<?>) {
+                if (a instanceof Reference ref) {
                     if (ref.valueType().id() == DataTypes.UNTYPED_OBJECT.id()) {
                         context.enforceThreeValuedLogic = true;
                         return null;
                     }
                 }
+                // For casts the second argument is always a null-literal carrying the target type
+                // Only visit the first argument to prevent visitLiteral on the second argument
+                // to set enforceThreeValuedLogic = true
+                boolean isNullable = context.isNullable;
```

Comment:
Noticed this due to query building failures after the change to use a `null` literal to carry the type info for implicit casts. I think as a side effect this will also cause it to build more optimized queries for cases where there is an explicit type cast, because now it will no longer enforce three-valued-logic for those.